medusa/restore_cluster.py

```diff
@@ -222,23 +224,13 @@ def _chunk(my_list, nb_chunks):
                 topology_matches = False
 
         if topology_matches:
-            # We can associate each restore node with exactly one backup node
-            backup_ringmap = collections.defaultdict(list)
-            target_ringmap = collections.defaultdict(list)
-            for token, host in backup_tokens.items():
-                backup_ringmap[token].append(host)
-            for token, host in target_tokens.items():
-                target_ringmap[token].append(host)
-
-            self.ringmap = backup_ringmap
-            i = 0
-            for token, hosts in backup_ringmap.items():
-                # take the node that has the same token list or pick the one with the same position in the map.
-                restore_host = target_ringmap.get(token, list(target_ringmap.values())[i])[0]
+            # backup and restore nodes are ordered by smallest token and associated one by one
+            sorted_backup_nodes = self._tokenmap_to_sorted_nodes(tokenmap)
+            sorted_target_nodes = self._tokenmap_to_sorted_nodes(target_tokenmap)
+            for i in range(len(sorted_backup_nodes)):
+                restore_host = sorted_target_nodes[i][0]
                 is_seed = True if self.fqdn_resolver.resolve_fqdn(restore_host) in self._get_seeds_fqdn() else False
-                self.host_map[restore_host] = {'source': [hosts[0]], 'seed': is_seed}
-                i += 1
-            logging.debug("self.host_map: {}".format(self.host_map))
+                self.host_map[restore_host] = {'source': [sorted_backup_nodes[i][0]], 'seed': is_seed}
```

Comment:
Is this safe in this context of k8ssandra? I ask because in the target cluster cass-operator is going to be specifying which nodes are seeds.

Reply:
Actually, and it was a happy surprise, it works perfectly. The intent of specifying which nodes are seeds is to allow Medusa to start the seed nodes first and then the rest of the nodes.
But this is only part of the restore orchestration part which we're not using in the context of k8ssandra. Which is fine, because cass-operator takes care of exactly that: seed nodes are labeled dynamically and nodes start one by one.